csharp/src/Google.Protobuf/WritingPrimitives.cs

```diff
@@ -163,10 +163,26 @@ public static void WriteBool(ref Span<byte> buffer, ref WriterInternalState stat
         /// </summary>
         public static void WriteString(ref Span<byte> buffer, ref WriterInternalState state, string value)
         {
-            // Optimise the case where we have enough space to write
-            // the string directly to the buffer, which should be common.
+            const int MaxBytesPerChar = 3;
+            const int MaxSmallStringLength = 128 / MaxBytesPerChar;
+
+            // The string is small enough that the length will always be a 1 byte varint.
+            // Also there is enough space to write length + bytes to buffer.
+            // Write string directly to the buffer, and then write length.
+            // This saves calling GetByteCount on the string. We get the string length from GetBytes.
+            if (value.Length <= MaxSmallStringLength && buffer.Length - state.position - 1 >= value.Length * MaxBytesPerChar)
+            {
+                // Get the original position, then increment it on state by 1, then write string to buffer.
+                // Method will return byte length, which is then set to the original position.
+                buffer[state.position++] = (byte)WriteStringToBuffer(buffer, ref state, value);
```

Comment:
I think  `int indexOfLengthDelimiter = state.position++;  buffer[indexOfLengthDelimiter] = WriteStringToBuffer(...)` would be easier to grasp and wouldn't require a comment?